From 8a0e6a4e47b982bfd6d1cce5e559b8173dadcd25 Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Mon, 22 May 2023 17:02:50 +0800
Subject: [PATCH 041/263] [Huawei] mm: mem_reliable: Update reliable page
 counter to zero if underflows

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I77BDW

--------------------------------

Since reliable page counter is used for debug purpose only, There is no
real function problem by doing this.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/mem_reliable.h | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/include/linux/mem_reliable.h b/include/linux/mem_reliable.h
index 437cd5c78db7..874afd991ddc 100644
--- a/include/linux/mem_reliable.h
+++ b/include/linux/mem_reliable.h
@@ -135,8 +135,19 @@ static inline bool reliable_allow_fb_enabled(void)
 static inline void reliable_page_counter(struct page *page,
 		struct mm_struct *mm, int val)
 {
-	if (page_reliable(page))
-		atomic_long_add(val, &mm->reliable_nr_page);
+	if (!page_reliable(page))
+		return;
+
+	atomic_long_add(val, &mm->reliable_nr_page);
+
+	/*
+	 * Update reliable page counter to zero if underflows.
+	 *
+	 * Since reliable page counter is used for debug purpose only,
+	 * there is no real function problem by doing this.
+	 */
+	if (unlikely(atomic_long_read(&mm->reliable_nr_page) < 0))
+		atomic_long_set(&mm->reliable_nr_page, 0);
 }
 
 static inline bool shmem_prepare_alloc(gfp_t *gfp_mask)
-- 
2.17.1

